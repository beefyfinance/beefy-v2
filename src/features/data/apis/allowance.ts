import { MultiCall, ShapeWithLabel } from 'eth-multicall';
import { AbiItem } from 'web3-utils';
import _vaultAbi from '../../../config/abi/vault.json';
import _boostAbi from '../../../config/abi/boost.json';
import _erc20Abi from '../../../config/abi/erc20.json';
import _multicallAbi from '../../../config/abi/multicall.json';
import Web3 from 'web3';
import { VaultEntity, VaultGov, VaultStandard } from '../entities/vault';
import { ChainEntity } from '../entities/chain';
import BigNumber from 'bignumber.js';
import { AllValuesAsString } from '../utils/types-utils';
import { BoostEntity } from '../entities/boost';
import { isTokenBoost, isTokenErc20, isTokenNative, TokenEntity } from '../entities/token';
import { BeefyState } from '../../redux/reducers';
import { selectTokenById } from '../selectors/tokens';

// fix TS typings
const boostAbi = _boostAbi as AbiItem[];
const erc20Abi = _erc20Abi as AbiItem[];
const multicallAbi = _multicallAbi as AbiItem[];

export interface VaultAllowance {
  vaultId: VaultEntity['id'];
  spenderAddress: string; // a 0x address
  allowance: BigNumber;
}

export interface BoostAllowance {
  boostId: BoostEntity['id'];
  spenderAddress: string; // a 0x address
  allowance: BigNumber;
}

/**
 * Get vault contract data
 */
export class AllowanceAPI {
  constructor(protected web3: Web3, protected chain: ChainEntity) {}

  /**
   * Do we really need this?
   */
  public async fetchGovVaultPoolAllowance(
    state: BeefyState,
    vaults: VaultGov[],
    walletAddress: string
  ): Promise<VaultAllowance[]> {
    const mc = new MultiCall(this.web3, this.chain.multicallAddress);

    const calls: ShapeWithLabel[] = [];
    for (const vault of vaults) {
      const token = selectTokenById(state, this.chain.id, vault.oracleId);
      if (!isTokenErc20(token)) {
        console.warn(`Token ${token.id} is not erc20, can't fetch allowance`);
        continue;
      }

      const tokenContract = new this.web3.eth.Contract(erc20Abi, token.contractAddress);
      calls.push({
        vaultId: vault.id, // not sure about this
        spenderAddress: vault.poolAddress,
        allowance: tokenContract.methods.allowance(walletAddress, vault.poolAddress),
      });
    }

    const [results] = (await mc.all([calls])) as AllValuesAsString<VaultAllowance>[][];

    // format strings as numbers
    return results.map(result => {
      return {
        vaultId: result.vaultId,
        spenderAddress: result.spenderAddress,
        allowance: new BigNumber(result.allowance || 0),
      } as VaultAllowance;
    });
  }

  public async fetchBoostAllowance(
    state: BeefyState,
    boosts: BoostEntity[],
    walletAddress: string
  ): Promise<BoostAllowance[]> {
    const mc = new MultiCall(this.web3, this.chain.multicallAddress);

    const calls: ShapeWithLabel[] = [];
    for (const boost of boosts) {
      const earnedToken = selectTokenById(state, this.chain.id, boost.earnedTokenId);
      if (!isTokenErc20(earnedToken)) {
        console.warn(`Token ${earnedToken.id} is not erc20, can't fetch allowance`);
        continue;
      }

      const tokenContract = new this.web3.eth.Contract(erc20Abi, earnedToken.contractAddress);
      calls.push({
        boostId: boost.id,
        spenderAddress: boost.earnContractAddress,
        allowance: tokenContract.methods.allowance(walletAddress, boost.earnContractAddress),
      });
    }

    const [results] = (await mc.all([calls])) as AllValuesAsString<BoostAllowance>[][];

    // format strings as numbers
    return results.map(result => {
      return {
        boostId: result.boostId,
        spenderAddress: result.spenderAddress,
        allowance: new BigNumber(result.allowance || 0),
      } as BoostAllowance;
    });
  }
}
