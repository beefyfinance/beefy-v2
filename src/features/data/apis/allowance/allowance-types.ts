import BigNumber from 'bignumber.js';
import { BeefyState } from '../../../redux/reducers/storev2';
import { BoostEntity } from '../../entities/boost';
import { TokenEntity } from '../../entities/token';
import { VaultEntity, VaultGov, VaultStandard } from '../../entities/vault';

export interface IAllowanceApi {
  fetchAllAllowances(
    state: BeefyState,
    standardVaults: VaultStandard[],
    govVaults: VaultGov[],
    boosts: BoostEntity[],
    walletAddress: string
  ): Promise<FetchAllAllowanceResult>;
}

export interface TokenAllowance {
  tokenId: TokenEntity['id'];
  spenderAddress: string; // a 0x address
  allowance: BigNumber;
}

export type FetchAllAllowanceResult = TokenAllowance[];

export type AllValuesAsString<T> = {
  [key in keyof T]: string;
};
