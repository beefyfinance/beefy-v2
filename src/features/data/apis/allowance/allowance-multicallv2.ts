import _BeefyV2AppMulticallUserAbi from '../../../../config/abi/BeefyV2AppUserMulticall.json';
import { AbiItem } from 'web3-utils';
import Web3 from 'web3';
import { VaultGov, VaultStandard } from '../../entities/vault';
import { ChainEntity } from '../../entities/chain';
import BigNumber from 'bignumber.js';
import { AllValuesAsString } from '../../utils/types-utils';
import { BoostEntity } from '../../entities/boost';
import { chunk } from 'lodash';
import { isTokenErc20, TokenEntity } from '../../entities/token';
import { FetchAllAllowanceResult, IAllowanceApi } from './allowance-types';
import { BeefyState } from '../../../redux/reducers/storev2';
import { selectTokenById } from '../../selectors/tokens';
import { featureFlag_getAllowanceApiChunkSize } from '../../utils/feature-flags';

// fix ts types
const BeefyV2AppMulticallUserAbi = _BeefyV2AppMulticallUserAbi as AbiItem | AbiItem[];

export class AllowanceMcV2API<T extends ChainEntity & { fetchBalancesAddress: string }>
  implements IAllowanceApi
{
  constructor(protected web3: Web3, protected chain: T) {}

  public async fetchAllAllowances(
    state: BeefyState,
    standardVaults: VaultStandard[],
    govVaults: VaultGov[],
    boosts: BoostEntity[],
    walletAddress: string
  ): Promise<FetchAllAllowanceResult> {
    const mc = new this.web3.eth.Contract(
      BeefyV2AppMulticallUserAbi,
      this.chain.fetchBalancesAddress
    );

    // first, build a list of tokens and spenders we want info on
    const allowanceCallsByToken: {
      [tokenAddress: string]: { tokenId: TokenEntity['id']; spenders: Set<string> };
    } = {};
    const addTokenIdToCalls = (tokenId: string, spenderAddress: string) => {
      const token = selectTokenById(state, this.chain.id, tokenId);
      if (!isTokenErc20(token)) {
        throw new Error("Can't query allowance of non erc20 token");
      }
      // TODO: temporary check until we can sort out the WFTM mystery
      if (!token.contractAddress) {
        console.error(`Could not find token contractAddress: ${token.id}`);
        return;
      }
      if (allowanceCallsByToken[token.contractAddress] === undefined) {
        allowanceCallsByToken[token.contractAddress] = { tokenId: token.id, spenders: new Set() };
      }
      allowanceCallsByToken[token.contractAddress].spenders.add(spenderAddress);
    };

    for (const standardVault of standardVaults) {
      addTokenIdToCalls(standardVault.earnedTokenId, standardVault.contractAddress);
      addTokenIdToCalls(standardVault.oracleId, standardVault.contractAddress);
    }
    for (const govVault of govVaults) {
      addTokenIdToCalls(govVault.oracleId, govVault.earnContractAddress);
    }
    for (const boost of boosts) {
      addTokenIdToCalls(boost.earnedTokenId, boost.earnContractAddress);
    }

    // if we send too much in a single call, we get "execution reversed"
    const CHUNK_SIZE = featureFlag_getAllowanceApiChunkSize();

    const allowanceCalls = Object.entries(allowanceCallsByToken);
    const callBatches = chunk(allowanceCalls, CHUNK_SIZE);

    const allowancePromises = callBatches.map(callBatch =>
      mc.methods
        .getAllowancesFlat(
          callBatch.map(([tokenAddress, _]) => tokenAddress),
          callBatch.map(([_, spendersCalls]) => Array.from(spendersCalls.spenders)),
          walletAddress
        )
        .call()
    );

    const results = (await Promise.all([...allowancePromises])) as AllValuesAsString<string[][]>;

    // now reasign results
    const res: FetchAllAllowanceResult = [];

    let resultsIdx = 0;

    for (const callBatch of callBatches) {
      const batchResults = results[resultsIdx];
      let resIdx = 0;
      for (const spendersCalls of callBatch.map(c => c[1])) {
        for (const spenderAddress of Array.from(spendersCalls.spenders)) {
          const allowance = batchResults[resIdx];
          if (allowance !== '0') {
            res.push({
              tokenId: spendersCalls.tokenId,
              spenderAddress,
              allowance: new BigNumber(allowance),
            });
          }
          resIdx++;
        }
      }
      resultsIdx++;
    }

    return res;
  }
}
