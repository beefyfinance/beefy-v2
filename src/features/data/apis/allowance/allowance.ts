import { MultiCall, ShapeWithLabel } from 'eth-multicall';
import { AbiItem } from 'web3-utils';
import _erc20Abi from '../../../../config/abi/erc20.json';
import Web3 from 'web3';
import { VaultGov, VaultStandard } from '../../entities/vault';
import { ChainEntity } from '../../entities/chain';
import BigNumber from 'bignumber.js';
import { AllValuesAsString } from '../../utils/types-utils';
import { BoostEntity } from '../../entities/boost';
import { isTokenErc20, TokenEntity, TokenErc20 } from '../../entities/token';
import { selectErc20TokenById, selectTokenById } from '../../selectors/tokens';
import { FetchAllAllowanceResult, IAllowanceApi, TokenAllowance } from './allowance-types';
import { BeefyState } from '../../../../redux-types';
import { createIdMap } from '../../utils/array-utils';
import { selectVaultById } from '../../selectors/vaults';

// fix TS typings
const erc20Abi = _erc20Abi as AbiItem[];

/**
 * Get vault contract data
 */
export class AllowanceAPI implements IAllowanceApi {
  constructor(protected web3: Web3, protected chain: ChainEntity) {}

  public async fetchAllAllowances(
    state: BeefyState,
    standardVaults: VaultStandard[],
    govVaults: VaultGov[],
    boosts: BoostEntity[],
    walletAddress: string
  ): Promise<FetchAllAllowanceResult> {
    // first, build a list of tokens and spenders we want info on
    const allowanceCallsByToken: {
      [tokenAddress: string]: { tokenId: TokenEntity['id']; spenders: Set<string> };
    } = {};
    const tokensById: { [tokenId: TokenEntity['id']]: TokenEntity } = {};
    const addTokenIdToCalls = (tokenId: string, spenderAddress: string) => {
      const token = selectTokenById(state, this.chain.id, tokenId);
      if (!isTokenErc20(token)) {
        throw new Error(`Can't query allowance of non erc20 token, skipping ${token.id}`);
      }
      if (allowanceCallsByToken[token.contractAddress] === undefined) {
        allowanceCallsByToken[token.contractAddress] = { tokenId: token.id, spenders: new Set() };
      }
      allowanceCallsByToken[token.contractAddress].spenders.add(spenderAddress);
      // keep a map to get decimals at the end
      if (tokensById[token.id] === undefined) {
        tokensById[token.id] = token;
      }
    };

    for (const standardVault of standardVaults) {
      addTokenIdToCalls(standardVault.earnedTokenId, standardVault.contractAddress);
      addTokenIdToCalls(standardVault.oracleId, standardVault.contractAddress);
      // special case for what seem to be a maxi vault
      const earnToken = selectErc20TokenById(state, this.chain.id, standardVault.earnedTokenId);
      addTokenIdToCalls(standardVault.oracleId, earnToken.contractAddress);
    }
    for (const govVault of govVaults) {
      addTokenIdToCalls(govVault.oracleId, govVault.earnContractAddress);
    }
    for (const boost of boosts) {
      const vault = selectVaultById(state, boost.vaultId);
      addTokenIdToCalls(vault.earnedTokenId, boost.earnContractAddress);
    }

    const calls: ShapeWithLabel[] = [];
    const baseTokenContract = new this.web3.eth.Contract(erc20Abi);
    for (const [tokenAddress, spendersCalls] of Object.entries(allowanceCallsByToken)) {
      const tokenContract = baseTokenContract.clone();
      tokenContract.options.address = tokenAddress;
      for (const spender of Array.from(spendersCalls.spenders)) {
        calls.push({
          tokenId: spendersCalls.tokenId, // not sure about this
          spenderAddress: spender,
          allowance: tokenContract.methods.allowance(walletAddress, spender),
        });
      }
    }

    type ResultType = AllValuesAsString<TokenAllowance>;
    const mc = new MultiCall(this.web3, this.chain.multicallAddress);
    const [results] = (await mc.all([calls])) as ResultType[][];

    return results.map(
      (result): TokenAllowance => ({
        tokenId: result.tokenId,
        spenderAddress: result.spenderAddress,
        allowance: new BigNumber(result.allowance).shiftedBy(-tokensById[result.tokenId].decimals),
      })
    );
  }

  async fetchTokensAllowance(tokens: TokenErc20[], walletAddress: string, spenderAddress: string) {
    const calls: ShapeWithLabel[] = [];
    const baseTokenContract = new this.web3.eth.Contract(erc20Abi);

    for (const token of tokens) {
      const tokenContract = baseTokenContract.clone();
      tokenContract.options.address = token.contractAddress;
      calls.push({
        tokenId: token.id,
        spenderAddress: spenderAddress,
        tokenAddress: token.contractAddress,
        allowance: tokenContract.methods.allowance(walletAddress, spenderAddress),
      });
    }

    type ResultType = AllValuesAsString<TokenAllowance>;
    const mc = new MultiCall(this.web3, this.chain.multicallAddress);
    const [results] = (await mc.all([calls])) as ResultType[][];

    const tokensById = createIdMap(tokens);

    return results.map(
      (result): TokenAllowance => ({
        tokenId: result.tokenId,
        spenderAddress: result.spenderAddress,
        allowance: new BigNumber(result.allowance).shiftedBy(-tokensById[result.tokenId].decimals),
      })
    );
  }
}
