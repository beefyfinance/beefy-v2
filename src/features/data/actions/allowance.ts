import { createAsyncThunk } from '@reduxjs/toolkit';
import type { BeefyState } from '../../../redux-types';
import type { FetchAllAllowanceResult } from '../apis/allowance/allowance-types';
import { getAllowanceApi } from '../apis/instances';
import type { ChainEntity } from '../entities/chain';
import type { TokenErc20 } from '../entities/token';
import type { VaultGov, VaultStandard } from '../entities/vault';
import { isGovVault } from '../entities/vault';
import { selectBoostById, selectBoostsByChainId } from '../selectors/boosts';
import { selectChainById } from '../selectors/chains';
import { selectVaultByChainId, selectVaultById } from '../selectors/vaults';
import { selectWalletAddress } from '../selectors/wallet';

interface ActionParams {
  chainId: ChainEntity['id'];
  walletAddress: string;
}

export interface FetchAllAllowanceFulfilledPayload {
  chainId: ChainEntity['id'];
  data: FetchAllAllowanceResult;
}

export const fetchAllAllowanceAction = createAsyncThunk<
  FetchAllAllowanceFulfilledPayload,
  ActionParams,
  { state: BeefyState }
>('allowance/fetchAllAllowanceAction', async ({ chainId, walletAddress }, { getState }) => {
  const state = getState();

  const userAddress = walletAddress || selectWalletAddress(state);
  const chain = selectChainById(state, chainId);
  const api = await getAllowanceApi(chain);

  // maybe have a way to retrieve those easily
  const boosts = selectBoostsByChainId(state, chainId).map(vaultId =>
    selectBoostById(state, vaultId)
  );
  const allVaults = selectVaultByChainId(state, chainId).map(vaultId =>
    selectVaultById(state, vaultId)
  );
  const standardVaults: VaultStandard[] = [];
  const govVaults: VaultGov[] = [];
  for (const vault of allVaults) {
    if (isGovVault(vault)) {
      govVaults.push(vault);
    } else {
      standardVaults.push(vault);
    }
  }

  // always re-fetch state as late as possible
  const data = await api.fetchAllAllowances(
    // always re-fetch state as late as possible
    getState(),
    standardVaults,
    govVaults,
    boosts,
    userAddress
  );
  return { chainId, data };
});

interface FetchAllowanceActionParams {
  chainId: ChainEntity['id'];
  spenderAddress: string;
  tokens: TokenErc20[];
  walletAddress: string;
}

export const fetchAllowanceAction = createAsyncThunk<
  FetchAllAllowanceFulfilledPayload,
  FetchAllowanceActionParams,
  { state: BeefyState }
>(
  'allowance/fetchAllowanceAction',
  async ({ chainId, spenderAddress, tokens, walletAddress }, { getState }) => {
    const state = getState();
    const userAddress = walletAddress || selectWalletAddress(state);
    const chain = selectChainById(state, chainId);
    const api = await getAllowanceApi(chain);

    const allowanceRes =
      userAddress && spenderAddress
        ? await api.fetchTokensAllowance(getState(), tokens, userAddress, spenderAddress)
        : [];

    return { chainId, data: allowanceRes };
  }
);
