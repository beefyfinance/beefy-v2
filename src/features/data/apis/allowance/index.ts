import { AllowanceAPI } from './allowance';
import { AllowanceMcV2API } from './allowance-multicallv2';

export { AllowanceAPI, AllowanceMcV2API };
