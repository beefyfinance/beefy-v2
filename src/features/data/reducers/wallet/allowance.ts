import { createSlice } from '@reduxjs/toolkit';
import type BigNumber from 'bignumber.js';
import { fetchAllAllowanceAction, fetchAllowanceAction } from '../../actions/allowance';
import type { TokenAllowance } from '../../apis/allowance/allowance-types';
import type { Draft } from 'immer';
import type { ChainEntity } from '../../entities/chain';
import type { TokenEntity } from '../../entities/token';
import { accountHasChanged, walletHasDisconnected } from './wallet';
import { selectVaultById } from '../../selectors/vaults';
import { reloadBalanceAndAllowanceAndGovRewardsAndBoostData } from '../../actions/tokens';
import { initiateBoostForm } from '../../actions/boosts';
import { selectBoostById } from '../../selectors/boosts';
import { initiateMinterForm } from '../../actions/minters';
import { selectMinterById } from '../../selectors/minters';

/**
 * State containing user allowances state
 * Allowance being the amount allowed to be spent big a contract
 * for the currently connected user
 */
export interface AllowanceState {
  byChainId: {
    [chainId in ChainEntity['id']]?: {
      byTokenAddress: {
        [tokenAddress: TokenEntity['address']]: {
          bySpenderAddress: {
            [spenderAddress: string]: BigNumber;
          };
        };
      };
    };
  };
}
export const initialAllowanceState: AllowanceState = { byChainId: {} };

export const allowanceSlice = createSlice({
  name: 'allowance',
  initialState: initialAllowanceState,
  reducers: {
    // standard reducer logic, with auto-generated action types per reducer
  },
  extraReducers: builder => {
    // reset state on user disconnect or address change
    builder.addCase(accountHasChanged, sliceState => {
      sliceState.byChainId = {};
    });
    builder.addCase(walletHasDisconnected, sliceState => {
      sliceState.byChainId = {};
    });

    builder.addCase(fetchAllAllowanceAction.fulfilled, (sliceState, action) => {
      const chainId = action.payload.chainId;
      const allowances = action.payload.data;
      addAllowancesToState(sliceState, chainId, allowances);
    });
    builder.addCase(fetchAllowanceAction.fulfilled, (sliceState, action) => {
      addAllowancesToState(sliceState, action.payload.chainId, action.payload.data);
    });
    builder.addCase(initiateBoostForm.fulfilled, (sliceState, action) => {
      const boost = selectBoostById(action.payload.state, action.payload.boostId);
      const vault = selectVaultById(action.payload.state, boost.vaultId);
      addAllowancesToState(sliceState, vault.chainId, action.payload.allowance);
    });
    builder.addCase(initiateMinterForm.fulfilled, (sliceState, action) => {
      const minter = selectMinterById(action.payload.state, action.payload.minterId);
      addAllowancesToState(sliceState, minter.chainId, action.payload.allowance);
    });

    builder.addCase(
      reloadBalanceAndAllowanceAndGovRewardsAndBoostData.fulfilled,
      (sliceState, action) => {
        const chainId = action.payload.chainId;
        const allowances = action.payload.allowance;
        addAllowancesToState(sliceState, chainId, allowances);
      }
    );
  },
});

function addAllowancesToState(
  sliceState: Draft<AllowanceState>,
  chainId: ChainEntity['id'],
  allowances: TokenAllowance[]
) {
  let stateForChain = sliceState.byChainId[chainId];
  if (stateForChain === undefined) {
    stateForChain = sliceState.byChainId[chainId] = { byTokenAddress: {} };
  }

  for (const tokenAllowance of allowances) {
    const tokenKey = tokenAllowance.tokenAddress.toLowerCase();
    let stateForToken = stateForChain.byTokenAddress[tokenKey];
    if (stateForToken === undefined) {
      stateForToken = stateForChain.byTokenAddress[tokenKey] = {
        bySpenderAddress: {},
      };
    }

    // only update data if necessary
    const spender = tokenAllowance.spenderAddress.toLowerCase();
    const stateForSpender = stateForToken.bySpenderAddress[spender];
    if (stateForSpender === undefined || !stateForSpender.isEqualTo(tokenAllowance.allowance)) {
      stateForToken.bySpenderAddress[spender] = tokenAllowance.allowance;
    }
  }
}
