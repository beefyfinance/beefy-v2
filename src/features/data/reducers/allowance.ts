import { createSlice } from '@reduxjs/toolkit';
import BigNumber from 'bignumber.js';
import { fetchAllAllowanceAction } from '../actions/allowance';
import { ChainEntity } from '../entities/chain';
import { TokenEntity } from '../entities/token';
import { accountHasChanged, walletHasDisconnected } from './wallet';

/**
 * State containing user allowances state
 * Allowance being the amount allowed to be spent big a contract
 * for the currently connected user
 */
export interface AllowanceState {
  byChainId: {
    [chainId: ChainEntity['id']]: {
      byTokenId: {
        [tokenId: TokenEntity['id']]: {
          bySpenderAddress: {
            [spenderAddress: string]: BigNumber;
          };
        };
      };
    };
  };
}
export const initialAllowanceState: AllowanceState = { byChainId: {} };

export const allowanceSlice = createSlice({
  name: 'allowance',
  initialState: initialAllowanceState,
  reducers: {
    // standard reducer logic, with auto-generated action types per reducer
  },
  extraReducers: builder => {
    // reset state on user disconnect or address change
    builder.addCase(accountHasChanged, sliceState => {
      sliceState.byChainId = {};
    });
    builder.addCase(walletHasDisconnected, sliceState => {
      sliceState.byChainId = {};
    });

    builder.addCase(fetchAllAllowanceAction.fulfilled, (sliceState, action) => {
      const chainId = action.payload.chainId;

      let stateForChain = sliceState.byChainId[chainId];
      if (stateForChain === undefined) {
        sliceState.byChainId[chainId] = { byTokenId: {} };
        stateForChain = sliceState.byChainId[chainId];
      }

      for (const tokenAllowance of action.payload.data) {
        let stateForToken = stateForChain.byTokenId[tokenAllowance.tokenId];
        if (stateForToken === undefined) {
          stateForChain.byTokenId[tokenAllowance.tokenId] = { bySpenderAddress: {} };
          stateForToken = stateForChain.byTokenId[tokenAllowance.tokenId];
        }

        // only update data if necessary
        let stateForSpender = stateForToken.bySpenderAddress[tokenAllowance.spenderAddress];
        if (stateForSpender === undefined || !stateForSpender.isEqualTo(tokenAllowance.allowance)) {
          stateForToken.bySpenderAddress[tokenAllowance.spenderAddress] = tokenAllowance.allowance;
        }
      }
    });
  },
});
