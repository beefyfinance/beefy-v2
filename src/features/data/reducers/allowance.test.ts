import BigNumber from 'bignumber.js';
import {
  fetchBoostAllowanceAction,
  FetchBoostAllowanceFulfilledPayload,
  fetchGovVaultPoolsAllowanceAction,
  FetchGovVaultPoolsAllowanceFulfilledPayload,
} from '../actions/allowance';
import { allowanceSlice, initialAllowanceState } from './allowance';

describe('Allowance slice tests', () => {
  it('should update state on fulfilled gov pool allowance', () => {
    const payload: FetchGovVaultPoolsAllowanceFulfilledPayload = {
      chainId: 'bsc',
      data: [
        { vaultId: 'banana-nfty-wbnb', spenderAddress: '0x000', allowance: new BigNumber(10) },
        { vaultId: 'belt-beltbnb', spenderAddress: '0x000', allowance: new BigNumber(5) },
      ],
    };
    const action = { type: fetchGovVaultPoolsAllowanceAction.fulfilled, payload: payload };
    const state = allowanceSlice.reducer(initialAllowanceState, action);
    expect(state).toMatchSnapshot();

    // getting the same vaults don't update the state object
    const beforeReDispatch = state.byChainId['bsc'].byVaultId;
    const newState = allowanceSlice.reducer(state, action);
    const afterReDispatch = newState.byChainId['bsc'].byVaultId;
    expect(beforeReDispatch).toBe(afterReDispatch);
  });

  it('should update state on fulfilled boost allowance', () => {
    const payload: FetchBoostAllowanceFulfilledPayload = {
      chainId: 'bsc',
      data: [
        {
          boostId: 'moo_ellipsis-renbtc-charge',
          allowance: new BigNumber(100),
          spenderAddress: '0x000',
        },
        {
          boostId: 'moo_beltbnb-czodiac',
          allowance: new BigNumber(100),
          spenderAddress: '0x000',
        },
      ],
    };
    const action = { type: fetchBoostAllowanceAction.fulfilled, payload: payload };
    const state = allowanceSlice.reducer(initialAllowanceState, action);
    expect(state).toMatchSnapshot();

    // getting the same vaults don't update the state object
    const beforeReDispatch = state.byChainId['bsc'].byBoostId;
    const newState = allowanceSlice.reducer(state, action);
    const afterReDispatch = newState.byChainId['bsc'].byBoostId;
    expect(beforeReDispatch).toBe(afterReDispatch);
  });
});
