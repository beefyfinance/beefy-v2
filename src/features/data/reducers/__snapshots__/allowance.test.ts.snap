// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Allowance slice tests should update state on fulfilled allowance 1`] = `
Object {
  "byChainId": Object {
    "bsc": Object {
      "byTokenId": Object {
        "token1": Object {
          "bySpenderAddress": Object {
            "0x0001": "10",
          },
        },
        "token2": Object {
          "bySpenderAddress": Object {
            "0x0001": "5",
            "0x0002": "5",
          },
        },
      },
    },
  },
}
`;
